File: 04-work-with-multiple-windows/java/src/test/java/MultipleWindows.java
package src.test.java;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.firefox.FirefoxDriver;
import java.util.Set;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.CoreMatchers.*;

public class MultipleWindows {
    WebDriver driver;

    @Before
    public void setUp() throws Exception {
        driver = new FirefoxDriver();
    }

    @After
    public void tearDown() throws Exception {
        driver.quit();
    }

    @Test
    public void multipleWindows() {
        driver.get("http://the-internet.herokuapp.com/windows");
        driver.findElement(By.cssSelector(".example a")).click();
        Object[] allWindows = driver.getWindowHandles().toArray();
        driver.switchTo().window(allWindows[0].toString());
        assertThat(driver.getTitle(), is(not("New Window")));
        driver.switchTo().window(allWindows[1].toString());
        assertThat(driver.getTitle(), is("New Window"));
    }

    @Test
    public void multipleWindowsRedux() {
        driver.get("http://the-internet.herokuapp.com/windows");

        // Get initial window handle
        String firstWindow = driver.getWindowHandle();
        // Create a newWindow variable
        String newWindow = "";
        // Trigger new window to open
        driver.findElement(By.cssSelector(".example a")).click();
        // Grab all window handles
        Set<String> allWindows = driver.getWindowHandles();

        // Iterate through window handles collection
        // Find the new window handle, storing it in the newWindow variable
        for (String window : allWindows) {
            if (!window.equals(firstWindow)) {
                newWindow = window;
            }
        }

        // Switch to the first window & verify
        driver.switchTo().window(firstWindow);
        assertThat(driver.getTitle(), is(not(equalTo("New Window"))));

        // Switch to the new window & verify
        driver.switchTo().window(newWindow);
        assertThat(driver.getTitle(), is(equalTo("New Window")));
    }

}

File: 04-work-with-multiple-windows/kotlin/src/test/kotlin/MultipleWindows.kt
import org.junit.Test
import org.openqa.selenium.By
import org.openqa.selenium.WebElement
import org.openqa.selenium.firefox.FirefoxDriver
import org.openqa.selenium.support.ui.ExpectedCondition
import org.openqa.selenium.support.ui.ExpectedConditions
import kotlin.test.AfterTest
import kotlin.test.assertEquals
import kotlin.test.assertNotEquals
import org.openqa.selenium.support.ui.WebDriverWait

class MultipleWindows {

    private val testUrl = "http://the-internet.herokuapp.com/windows"
    private val firstTabTitle = "The Internet"
    private val secondTabTitle = "New Window"

    private val driver by lazy {
        FirefoxDriver()
    }

    private val wait by lazy {
        WebDriverWait(driver, 20)
    }

    @AfterTest
    fun tearDown() {
        // Quit session (closes browser)
        driver.quit()
    }

    @Test
    fun multipleWindows() {
        driver.get(testUrl)
        driver.findElement(By.cssSelector(".example a")).click()

        val allWindows = driver.windowHandles.toTypedArray()

        driver.switchTo().window(allWindows[0].toString())
        wait.until(ExpectedConditions.titleIs(firstTabTitle))
        assertNotEquals(
            secondTabTitle,
            driver.title
        )


        driver.switchTo().window(allWindows[1].toString())
        wait.until(ExpectedConditions.titleIs(secondTabTitle))
        assertEquals(
            secondTabTitle,
            driver.title
        )
    }

    @Test
    fun multipleWindowsRedux() {
        driver.get(testUrl)

        // Get initial window handle
        val firstWindow = driver.windowHandle

        // Trigger new window to open
        driver.findElement(By.cssSelector(".example a")).click()

        // Grab all window handles
        val allWindows = driver.windowHandles

        // Iterate through window handles collection
        // Find the new window handle, storing it in the newWindow variable
        for (window in allWindows) {
            if (window != firstWindow) {
                // Switch to the new window & verify
                driver.switchTo().window(window)
                wait.until(ExpectedConditions.titleIs(secondTabTitle))
                assertEquals(
                    secondTabTitle,
                    driver.title
                )
            }
        }

        // Switch to the first window & verify
        driver.switchTo().window(firstWindow)
        wait.until(ExpectedConditions.titleIs(firstTabTitle))
        assertNotEquals(
            secondTabTitle,
            driver.title
        )
    }
}